### PR TITLE
Fix emitting enums with string keys

### PR DESCRIPTION
**Pull request by deadalusai**
This PR fixes the routine which processes enum types to work with enums with string keys. Additionally the routine now emits a reverse mapping for all "numeric" enum members, [matching the behavior of the TypeScript compiler more closely](https://www.typescriptlang.org/play/#src=%2F%2F%20String%20valued%20enum%2C%20no%20reverse%20mapping%0D%0Aenum%20StringEnum%20%7B%0D%0A%20%20STR%20%3D%20'abc'%2C%0D%0A%20%20OTHER_STR%20%3D%20'xyz'%2C%20%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Number%20valued%20enum%2C%20with%20reverse%20mapping%0D%0Aenum%20StringKeyEnum%20%7B%0D%0A%20%20'string%20key'%20%3D%201%2C%0D%0A%20%20'string%20key%202'%20%3D%200%2C%0D%0A%7D%0D%0A%0D%0A%2F%2F%20String%2Fnumber%20mixed%20enum%2C%20reverse%20mapping%20for%20numbers%20only%0D%0Aenum%20MixedEnum%20%7B%0D%0A%20%20STR%20%3D%20'abc'%2C%0D%0A%20%20NUM%20%3D%203%2C%0D%0A%20%20'string%20key'%20%3D%200%2C%0D%0A%7D%0D%0A).

### CODE REVIEW

File: src/tsickle.ts [hunk 1]
```diff
@@ -1542,11 +1542,20 @@ class Annotator extends ClosureRewriter {
...
+      // Initializer expressions are not allowed for string enum members so this routine
+      // assumes that any enum members with an initializer have type number.
+      if (typeof memberValue === 'string') {
+        // String valued members are not reverse mapped.
+        continue;
+      }
+      if (member.name.kind === ts.SyntaxKind.Identifier) {
         this.emit(`${name}[${name}.${memberName}] = "${memberName}";\n`);
+      } else {
+        this.emit(`${name}[${name}[${memberName}]] = ${memberName};\n`);
```
Comment (mprobst) [inline src/tsickle.ts]: Can you add a comment to this else branch explaining what it covers? (expression keys)

File: src/tsickle.ts [hunk 2]
```diff
@@ -1542,11 +1542,20 @@ class Annotator extends ClosureRewriter {
...
-    if (enumType === 'number') {
-      for (const member of node.members) {
-        const memberName = member.name.getText();
+    // Emit the reverse mapping of foo[foo.BAR] = 'BAR'; lines for number enum members
+    for (const member of node.members) {
+      const memberName = member.name.getText();
+      const memberValue = this.typeChecker.getConstantValue(member);
+      // Initializer expressions are not allowed for string enum members so this routine
+      // assumes that any enum members with an initializer have type number.
+      if (typeof memberValue === 'string') {
```
Comment (mprobst) [inline src/tsickle.ts]: should we invert this and rather be conservative, i.e. check for `typeof memberValue !== 'number'`?

File: test_files/enum/enum.ts [hunk 3]
```diff
@@ -61,7 +61,13 @@ enum StringEnum {
   OTHER_STR = 'xyz',
 }
 
+enum StringKeyEnum {
```
Comment (mprobst) [inline test_files/enum/enum.ts]: can you add comments to the enums explaining exactly what case this is testing?

File: src/tsickle.ts (hunk 2 above)
Comment (deadalusai) [reply src/tsickle.ts]: 👍

File: test_files/enum/enum.ts (hunk 3 above)
Comment (deadalusai) [reply test_files/enum/enum.ts]: 👍

**Comment by deadalusai** [comment]
Thanks Martin. I'm working on getting the CLA signed through my company and will rebase as soon as I have that in place :)

**Comment by deadalusai** [comment]
@googlebot I've signed the CLA

**Comment by deadalusai** [comment]
I signed it!

**Comment by deadalusai** [comment]
I signed it!

**Comment by deadalusai** [comment]
@mprobst Apologies for the comment spam - trying to trigger the googlebot. I've rebased onto master and squashed as of yesterday. Let me know if you'd like another rebase.

**Comment by mprobst** [comment]
@deadalusai you haven't lived until you've had a two hour conversation with @googlebot. I'll see to merge this in.

**Comment by evmar** [comment]
This fails in google3:

third_party/javascript/node_modules/tsickle/src/tsickle.ts(1439,40): error TS2694: Namespace '"/build/work/8fc805ae8e082f30905d8fa4c6832817/google3/third_party/javascript/node_modules/tsickle/src/typescript"' has no exported member 'EnumMember'.

1439   private getEnumMemberType(member: ts.EnumMember): 'number'|'string' {

Something about TypeScript versions maybe?

**Comment by deadalusai** [comment]
@evmar Tsickle re-exports everything it uses from TypeScript in `src/typescript.ts`. Check out this line in the diff: https://github.com/angular/tsickle/pull/745/files#diff-0ad76a5997509089412e04d8a385052cL23

I'm not sure what your source looks like - perhaps the change to this line was lost in a merge?

**Comment by evmar** [comment]
I wonder if maybe this is a TS2.7 API:
https://github.com/angular/tsickle/blob/a60b629a6282f74fcc8bfeaf63e2a9c262045ea1/yarn.lock#L2124

and in Google we use TS2.6.

**Comment by deadalusai** [comment]
TS2.6 does export `EnumMember`: https://github.com/Microsoft/TypeScript/blob/v2.6.0/src/compiler/types.ts#L1927

**Comment by evmar** [comment]
Sorry for the noise, Martin pointed me at the Google-internal place we need to update for this change.

**Pull request merged**